### DISCUSSION TITLE
Automatic multithreading systems and resources/queries

### DISCUSSION THREAD

**Discussion by Tonaion02** [bevyengine/bevy#14994] [Q&A]
Hello guys, i have a question.
I know that if you put in arguments of a function the resources to retrieve and the queries to make, the systems are automatically parallelized. 
I know that you can retrieve resources or make query in the corpe of a function, but in the case where i do in that way, what happens to automatic multithreading between systems? It doesn't work anymore?

**Comment by SkiFire13** [comment]
> I know that you can retrieve resources or make query in the corpe of a function

Do you mean when using a `&mut World` parameter? In that case while your system is running no other system can, since your system is the only one that can access any part of the world.

**Reply by Tonaion02** [reply]
Yeah, i think that is the only way to retrieve resources or make queries in the corpe of the function, am i right?

In the case this is true, you can indicate to me a piece of documentation where is written that using world cause that this system can't run in parallel with other systems?

**Reply by SkiFire13** [reply]
>Yeah, i think that is the only way to retrieve resources or make queries in the corpe of the function, am i right?

Other than requiring them as parameters of the system yes.

> In the case this is true, you can indicate to me a piece of documentation where is written that using world cause that this system can't run in parallel with other systems?

I don't think it is documented what can run in parallel with what, it's more like an implementation detail.

In practice though Bevy must be conservative, if a system could potentially access something then the scheduler must consider the case where it will access it. A system taking `&mut World` as input could access anything, so it must conflict with any other system.

**Reply by Tonaion02** [reply]
Ok, thx for the help. And commands? If the commands are not applied immediately i think that you can use commands and the systems can run in parallel, am i wrong?

**Reply by SkiFire13** [reply]
Yes, a `Commands` parameter won't prevent your system from being run in parallel with other systems.

However note that the phase when the commands are applied later on will not run in parallel because it takes a `&mut World`.

**Reply by Tonaion02** [reply]
Thx a lot!

**Reply by Tonaion02** [reply]
Hello, i have another doubt, if i have multiple systems with a `Commands` parameter, they can run in parallel? The documentation suggested that `Commands` is really a queue, so probably can run in parallel, am i wrong?

**Reply by SkiFire13** [reply]
Yes, two systems with `Commands` can run in parallel (of course the other parameters must also be compatible with each other). Each system has its own queue, so they don't conflict on that (as opposed to for example `EventWriter`, which writes to a queue but that's a in a resource and so it's shared between all systems)

**Reply by Tonaion02** [reply]
Ok thx a lot!
